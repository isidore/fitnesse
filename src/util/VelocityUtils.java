package util;

import java.io.IOException;
import java.io.StringWriter;

import org.apache.velocity.Template;
import org.apache.velocity.VelocityContext;
import org.apache.velocity.util.ContextAware;

import com.spun.util.parser.ParserCommons;

import fitnesse.VelocityFactory;

public class VelocityUtils {

  public static String parseTemplate(String velocityTemplate, ContextAware aware)
      throws Exception, IOException {
    Template template = VelocityFactory.getVelocityEngine().getTemplate(
        velocityTemplate);
    StringWriter writer = new StringWriter();
    VelocityContext velocityContext = new VelocityContext();
    aware.setContext(velocityContext);
    velocityContext.put("common", new ParserCommons());

    template.merge(velocityContext, writer);
    return writer.toString();
  }

}
