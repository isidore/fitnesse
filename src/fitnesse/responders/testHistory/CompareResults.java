package fitnesse.responders.testHistory;

import java.util.Date;
import java.util.List;

public class CompareResults {

  private final Comparison comparison;
  private final List<String> firstTableResults;
  private final List<String> secondTableResults;
  private final List<String> resultContent;
  private final List<MatchedPair> matchedTables;
  private Date firstTableCreationDate;
  private Date secondTableCreationDate;
  private boolean completeMatch;

  public enum Comparison {
    Uncomparable, Comparable

  }

  public CompareResults(Comparison comparison) {
    this(comparison, null, null, null, null);

  }

  public Date getFirstTableCreationDate() {
    return firstTableCreationDate;
  }

  public Date getSecondTableCreationDate() {
    return secondTableCreationDate;
  }

  public CompareResults(Comparison comparison, List<String> firstTableResults,
      List<String> secondTableResults, List<String> resultContent,
      List<MatchedPair> matchedTables) {
    this.comparison = comparison;
    this.firstTableResults = firstTableResults;
    this.secondTableResults = secondTableResults;
    this.resultContent = resultContent;
    this.matchedTables = matchedTables;
  }

  public boolean isComparisonPossible() {
    return comparison == Comparison.Comparable;
  }

  public List<String> getFirstTableResults() {
    return firstTableResults;
  }

  public List<String> getSecondTableResults() {
    return secondTableResults;
  }

  public List<String> getResultContent() {
    return resultContent;
  }

  public MatchedPair findMatchByFirstTableIndex(int firstIndex) {
    return MatchedPair.findMatchByFirstTableIndex(matchedTables, firstIndex);
  }

  public void setFirstFileDate(Date firstTableCreationDate) {
    this.firstTableCreationDate = firstTableCreationDate;
  }

  public void setSecondFileDate(Date secondTableCreationDate) {
    this.secondTableCreationDate = secondTableCreationDate;
  }

  public boolean isCompleteMatch() {
    return completeMatch;
  }

  public void setCompleteMatch(boolean completeMatch) {
    this.completeMatch = completeMatch;
  }

}
