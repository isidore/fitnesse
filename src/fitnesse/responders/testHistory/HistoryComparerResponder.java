package fitnesse.responders.testHistory;

import java.io.File;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Set;

import org.apache.velocity.context.Context;
import org.apache.velocity.util.ContextAware;

import util.VelocityUtils;
import fitnesse.FitNesseContext;
import fitnesse.Responder;
import fitnesse.http.Request;
import fitnesse.http.Response;
import fitnesse.http.SimpleResponse;
import fitnesse.responders.ErrorResponder;
import fitnesse.responders.templateUtilities.PageTitle;
import fitnesse.wiki.PathParser;

public class HistoryComparerResponder implements Responder, ContextAware {
  public static class FileExister {
    public boolean filesExist(String... files) {
      for (String file : files) {
        if (!new File(file).exists()) {
          return false;
        }
      }
      return true;
    }
  }

  private static final String COMPARE_HISTORY_TEMPLATE = "compareHistory.vm";
  private static SimpleDateFormat DATE_FORMAT = new SimpleDateFormat(
      TestHistory.TEST_RESULT_FILE_DATE_PATTERN);

  private HistoryComparer comparer;
  private PageTitle pageTitle;
  private CompareResults compareResult;
  FileExister fileExister = new FileExister();

  public HistoryComparerResponder(HistoryComparer historyComparer) {
    comparer = historyComparer;
  }

  public HistoryComparerResponder() {
    comparer = new HistoryComparer();
  }

  public Response makeResponse(FitNesseContext context, Request request)
      throws Exception {
    pageTitle = new PageTitle("Test History", PathParser.parse(request
        .getResource()));
    List<String> fileNames = getFileNamesFromRequest(request);
    if (fileNames.size() != 2)
      return new ErrorResponder(
          ("Compare Failed because the wrong number of Input Files were given. "
              + "Select two please.")).makeResponse(context, request);
    String firstFilePath = composeFileName(request, fileNames.get(0), context);
    String secondFilePath = composeFileName(request, fileNames.get(1), context);

    if (!fileExister.filesExist(firstFilePath, secondFilePath))
      return new ErrorResponder(
          "Compare Failed because the files were not found.").makeResponse(
          context, request);

    return makeResponseFromComparison(context, request, fileNames,
        firstFilePath, secondFilePath);
  }

  private Response makeResponseFromComparison(FitNesseContext context,
      Request request, List<String> fileNames, String firstFilePath,
      String secondFilePath) throws Exception {
    compareResult = comparer.compare(firstFilePath, secondFilePath);
    compareResult.setFirstFileDate(parseDateFromFile(fileNames.get(0)));
    compareResult.setSecondFileDate(parseDateFromFile(fileNames.get(1)));
    return renderResponse(context, request, compareResult);
  }

  public Response renderResponse(FitNesseContext context, Request request,
      CompareResults compareResult) throws Exception {
    this.compareResult = compareResult;
    if (compareResult.isComparisonPossible())
      return makeValidResponse();
    else {
      String message = String.format("These files could not be compared."
          + "  They might be suites, or something else might be wrong.");
      return new ErrorResponder(message).makeResponse(context, request);
    }
  }

  private static Date parseDateFromFile(String fileName) {
    try {
      return DATE_FORMAT.parse(fileName);
    } catch (ParseException e) {
      return null;
    }
  }

  private String composeFileName(Request request, String fileName,
      FitNesseContext context) {
    return context.getTestHistoryDirectory().getPath() + File.separator
        + request.getResource() + File.separator + fileName;
  }

  private List<String> getFileNamesFromRequest(Request request) {
    Set<String> keys = request.getMap().keySet();
    ArrayList<String> files = new ArrayList<String>();
    for (String key : keys) {
      if (key.contains(Request.TEST_RESULT_PREFIX)) {
        files.add(key.substring(key.indexOf("_") + 1));
      }
    }
    return files;
  }

  private Response makeValidResponse() throws Exception {
    String text = VelocityUtils.parseTemplate(COMPARE_HISTORY_TEMPLATE, this);
    return makeResponseFromTemplate(text);

  }

  private Response makeResponseFromTemplate(String text) throws Exception {
    SimpleResponse response = new SimpleResponse();
    response.setContent(text);
    return response;
  }

  public void setTestComparer(HistoryComparer comparer) {
    this.comparer = comparer;
  }

  @Override
  public void setContext(Context velocityContext) {

    velocityContext.put("results", compareResult);
    velocityContext.put("pageTitle", pageTitle);

  }
}
