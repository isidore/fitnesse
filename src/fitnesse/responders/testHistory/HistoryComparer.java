package fitnesse.responders.testHistory;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

import org.htmlparser.util.ParserException;

import fitnesse.responders.run.TestExecutionReport;
import fitnesse.responders.testHistory.CompareResults.Comparison;
import fitnesse.slimTables.HtmlTableScanner;

public class HistoryComparer {
  public static class ComparerResults {

    public static final String NO_MATCH = "fail";
    public static final String MATCH = "pass";
  }

  private TableListComparer comparer;

  public String secondFileContent = "";
  public String firstFileContent = "";
  public File resultFile;
  public ArrayList<String> resultContent = new ArrayList<String>();
  public HtmlTableScanner firstScanner;
  public HtmlTableScanner secondScanner;
  public List<String> firstTableResults;
  public List<String> secondTableResults;
  public ArrayList<MatchedPair> matchedTables;
  private static final String blankTable = "<table><tr><td></td></tr></table>";

  public String getFileContent(String filePath) {
    TestExecutionReport report;
    try {
      report = new TestExecutionReport().read(new File(filePath));
      if (report.getResults().size() != 1)
        return null;
      return report.getResults().get(0).getContent();
    } catch (Exception e) {
      throw new RuntimeException();
    }
  }

  public boolean allTablesMatch() {
    if (matchedTables == null || matchedTables.size() == 0
        || firstTableResults == null || firstTableResults.size() == 0)
      return false;
    if (matchedTables.size() == firstTableResults.size()) {
      for (MatchedPair match : matchedTables) {
        if (match.matchScore < (MatchedPair.MAX_MATCH_SCORE - .01))
          return false;
      }
      return true;
    }
    return false;
  }

  public CompareResults compare(String firstFilePath, String secondFilePath)
      throws Exception {
    if (firstFilePath.equals(secondFilePath))
      return new CompareResults(Comparison.Uncomparable);
    initializeFileContents(firstFilePath, secondFilePath);
    return grabAndCompareTablesFromHtml();
  }

  public CompareResults grabAndCompareTablesFromHtml() throws ParserException {
    initializeComparerHelpers();
    if (firstScanner.getTableCount() == 0 || secondScanner.getTableCount() == 0)
      return new CompareResults(Comparison.Uncomparable);
    comparer = new TableListComparer(firstScanner, secondScanner);
    comparer.compareAllTables();
    matchedTables = comparer.tableMatches;
    getTableTextFromScanners();
    lineUpTheTables();
    addBlanksToUnmatchingRows();
    makePassFailResultsFromMatches();
    CompareResults compareResults = new CompareResults(Comparison.Comparable,
        firstTableResults, secondTableResults, resultContent, matchedTables);
    compareResults.setCompleteMatch(allTablesMatch());
    return compareResults;
  }

  private void initializeComparerHelpers() throws ParserException {
    matchedTables = new ArrayList<MatchedPair>();
    firstScanner = new HtmlTableScanner(firstFileContent);
    secondScanner = new HtmlTableScanner(secondFileContent);
  }

  public void lineUpTheTables() {
    for (int currentMatch = 0; currentMatch < matchedTables.size(); currentMatch++)
      lineUpMatch(currentMatch);
    lineUpLastRow();

  }

  private void lineUpMatch(int currentMatch) {
    insertBlanksUntilMatchLinesUp(new FirstResultAdjustmentStrategy(),
        currentMatch);
    insertBlanksUntilMatchLinesUp(new SecondResultAdjustmentStrategy(),
        currentMatch);
  }

  private void insertBlanksUntilMatchLinesUp(
      ResultAdjustmentStrategy adjustmentStrategy, int currentMatch) {
    while (adjustmentStrategy.matchIsNotLinedUp(currentMatch)) {
      adjustmentStrategy.insertBlankTableBefore(currentMatch);
      incrementRemaingMatchesToCompensateForInsertion(adjustmentStrategy,
          currentMatch);
    }
  }

  private void incrementRemaingMatchesToCompensateForInsertion(
      ResultAdjustmentStrategy adjustmentStrategy, int currentMatch) {
    for (int matchToAdjust = currentMatch; matchToAdjust < matchedTables.size(); matchToAdjust++) {
      matchedTables.set(matchToAdjust, adjustmentStrategy
          .getAdjustedMatch(matchToAdjust));
    }
  }

  private interface ResultAdjustmentStrategy {
    boolean matchIsNotLinedUp(int matchIndex);

    void insertBlankTableBefore(int matchIndex);

    MatchedPair getAdjustedMatch(int matchIndex);
  }

  private class FirstResultAdjustmentStrategy implements
      ResultAdjustmentStrategy {
    public boolean matchIsNotLinedUp(int matchIndex) {
      MatchedPair matchedPair = matchedTables.get(matchIndex);
      return matchedPair.first < matchedPair.second;
    }

    public void insertBlankTableBefore(int matchIndex) {
      firstTableResults.add(matchedTables.get(matchIndex).first, blankTable);
    }

    public MatchedPair getAdjustedMatch(int matchIndex) {
      MatchedPair matchedPair = matchedTables.get(matchIndex);
      return new MatchedPair(matchedPair.first + 1, matchedPair.second,
          matchedPair.matchScore);

    }
  }

  private class SecondResultAdjustmentStrategy implements
      ResultAdjustmentStrategy {
    public boolean matchIsNotLinedUp(int matchIndex) {
      MatchedPair matchedPair = matchedTables.get(matchIndex);
      return matchedPair.first > matchedPair.second;
    }

    public void insertBlankTableBefore(int matchIndex) {
      secondTableResults.add(matchedTables.get(matchIndex).second, blankTable);
    }

    public MatchedPair getAdjustedMatch(int matchIndex) {
      MatchedPair matchedPair = matchedTables.get(matchIndex);
      return new MatchedPair(matchedPair.first, matchedPair.second + 1,
          matchedPair.matchScore);
    }
  }

  private void lineUpLastRow() {
    while (firstTableResults.size() > secondTableResults.size())
      secondTableResults.add(blankTable);
    while (secondTableResults.size() > firstTableResults.size())
      firstTableResults.add(blankTable);
  }

  public void addBlanksToUnmatchingRows() {
    for (int tableIndex = 0; tableIndex < firstTableResults.size(); tableIndex++) {
      if (tablesDontMatchAndArentBlank(tableIndex)) {
        insetBlanksToSplitTheRow(tableIndex);
        incrementMatchedPairsIfBelowTheInsertedBlank(tableIndex);
      }
    }
  }

  private boolean tablesDontMatchAndArentBlank(int tableIndex) {
    return !thereIsAMatchForTableWithIndex(tableIndex)
        && firstAndSecondTableAreNotBlank(tableIndex);
  }

  private boolean thereIsAMatchForTableWithIndex(int tableIndex) {
    return MatchedPair.findMatchByFirstTableIndex(matchedTables, tableIndex)
        .getScore() > 0.1;
  }

  private boolean firstAndSecondTableAreNotBlank(int tableIndex) {
    return !(firstTableResults.get(tableIndex).equals(blankTable) || secondTableResults
        .get(tableIndex).equals(blankTable));
  }

  private void incrementMatchedPairsIfBelowTheInsertedBlank(int tableIndex) {
    for (int j = 0; j < matchedTables.size(); j++) {
      MatchedPair match = matchedTables.get(j);
      if (match.first > tableIndex)
        matchedTables.set(j, new MatchedPair(match.first + 1, match.second + 1,
            match.matchScore));
    }
  }

  private void insetBlanksToSplitTheRow(int tableIndex) {
    secondTableResults.add(tableIndex, blankTable);
    firstTableResults.add(tableIndex + 1, blankTable);
  }

  private void getTableTextFromScanners() {
    firstTableResults = new ArrayList<String>();
    secondTableResults = new ArrayList<String>();
    for (int i = 0; i < firstScanner.getTableCount(); i++)
      firstTableResults.add(firstScanner.getTable(i).toHtml());

    for (int i = 0; i < secondScanner.getTableCount(); i++)
      secondTableResults.add(secondScanner.getTable(i).toHtml());
  }

  public void makePassFailResultsFromMatches() {
    for (int i = 0; i < firstTableResults.size(); i++) {
      String result = ComparerResults.NO_MATCH;
      for (MatchedPair match : matchedTables) {
        if (match.first == i && match.matchScore >= 1.19) {
          result = ComparerResults.MATCH;
        }
      }
      resultContent.add(result);

    }
  }

  private void initializeFileContents(String firstFilePath,
      String secondFilePath) throws ParserException {
    String content = getFileContent(firstFilePath);
    firstFileContent = content == null ? "" : content;
    content = getFileContent(secondFilePath);
    secondFileContent = content == null ? "" : content;
  }

  public List<String> getResultContent() {
    return resultContent;
  }

}
