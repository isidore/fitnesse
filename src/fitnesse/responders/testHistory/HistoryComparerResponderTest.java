package fitnesse.responders.testHistory;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;
import junit.framework.TestCase;

import org.approvaltests.Approvals;
import org.approvaltests.UseReporter;
import org.approvaltests.reporters.FailedDiffTextApprovalReport;

import util.FileUtil;
import util.ListUtility;
import fitnesse.http.Request;
import fitnesse.http.SettableRequest;
import fitnesse.http.SimpleResponse;
import fitnesse.http.Response.ReasonCodes;
import fitnesse.responders.testHistory.CompareResults.Comparison;
import fitnesse.responders.testHistory.HistoryComparer.ComparerResults;
import fitnesse.responders.testHistory.HistoryComparerResponder.FileExister;
import fitnesse.testutil.FitNesseUtil;

@UseReporter(FailedDiffTextApprovalReport.class)
public class HistoryComparerResponderTest extends TestCase {

  public void testShouldBeAbleToCompareTwoHistories() throws Exception {
    String content = "<table><tr><td>This is the content</td></tr></table>";
    CompareResults results = new CompareResults(Comparison.Comparable,
        ListUtility.list(content), ListUtility.list(content), ListUtility
            .list(ComparerResults.MATCH), ListUtility.asList(new MatchedPair(0,
            0, .1)));
    results.setCompleteMatch(false);

    SimpleResponse response = runForResults(results, true);
    assertEquals(ReasonCodes.OK, response.getStatus());
    Approvals.approveHtml(response.getContent());
  }

  public void testShouldReturnErrorPageIfCompareFails() throws Exception {
    CompareResults results = new CompareResults(Comparison.Uncomparable);
    SimpleResponse response = runForResults(results, true);
    assertEquals(400, response.getStatus());
    Approvals.approveHtml(response.getContent());
  }

  public void testShouldReturnErrorPageIfFilesDontExist() throws Exception {
    SimpleResponse response = runForResults(null, false);
    assertEquals(400, response.getStatus());
    Approvals.approveHtml(response.getContent());
  }

  public void testShouldReturnErrorPageIfOnly1InputFile() throws Exception {
    SettableRequest request = new SettableRequest();
    request.addInput(Request.TEST_RESULT_PREFIX + "20090215111430.txt", "");
    request.setResource("TestFolder");

    HistoryComparerResponder responder = new HistoryComparerResponder();
    SimpleResponse response = (SimpleResponse) responder.makeResponse(
        FitNesseUtil.makeTestContext(null), request);
    
    assertEquals(400, response.getStatus());
    Approvals.approveHtml(response.getContent());
  }

  private SimpleResponse runForResults(CompareResults results,
      boolean filesExist) throws Exception {
    HistoryComparerResponder responder = new HistoryComparerResponder();
    responder.fileExister = new FilesExistMock(filesExist);
    SimpleResponse response = (SimpleResponse) responder.makeResponse(
        FitNesseUtil.makeTestContext(null), setupRequest(responder, results));
    return response;
  }

  private Request setupRequest(HistoryComparerResponder responder,
      CompareResults results) throws Exception {
    String folder = "TestFolder";
    String basePath = FileUtil
        .useCorrectPathSeperator("./TestDir/files/testResults/" + folder + "/");
    String name1 = "20090215111430.txt";
    String name2 = "20090215111435.txt";
    HistoryComparer mockedComparer = mockOutComparer(results, basePath + name1,
        basePath + name2);
    responder.setTestComparer(mockedComparer);

    SettableRequest request = createCompareRequestForFiles(folder, name1, name2);

    return request;
  }

  private SettableRequest createCompareRequestForFiles(String folder,
      String name1, String name2) {
    SettableRequest request = new SettableRequest();
    request.addInput(Request.TEST_RESULT_PREFIX + name1, "");
    request.addInput(Request.TEST_RESULT_PREFIX + name2, "");
    request.setResource(folder);
    return request;
  }

  private HistoryComparer mockOutComparer(CompareResults results,
      String firstFile, String secondFile) throws Exception {

    HistoryComparer mockedComparer = mock(HistoryComparer.class);
    when(mockedComparer.compare(firstFile, secondFile)).thenReturn(results);
    return mockedComparer;
  }

  /* Inner Classes */

  public static class FilesExistMock extends FileExister {
    private final boolean exists;

    public FilesExistMock(boolean exists) {
      this.exists = exists;
    }

    @Override
    public boolean filesExist(String... files) {
      return exists;
    }
  }

}
