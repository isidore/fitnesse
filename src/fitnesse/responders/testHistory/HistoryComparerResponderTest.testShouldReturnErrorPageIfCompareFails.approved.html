<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html>
	<head>
		<title>Error Occured</title>
		<link rel="stylesheet" type="text/css" href="/files/css/fitnesse.css" media="screen"/>
		<link rel="stylesheet" type="text/css" href="/files/css/fitnesse_print.css" media="print"/>
		<script src="/files/javascript/fitnesse.js" type="text/javascript"></script>
	</head>
	<body>
		<div class="sidebar">
			<div class="art_niche" onclick="document.location='FrontPage'"></div>
			<div class="actions"></div>
		</div>
		<div class="mainbar">
			<div class="header">
				<span class="page_title">Error Occured</span>
			</div>
			<div class="main">
				<div class="centered">These files could not be compared.  They might be suites, or something else might be wrong.</div>
			</div>
		</div>
	</body>
</html>
